### PR TITLE
Revert "Revert "Unify meta tensor and fake tensor converter conversion (#87943)""

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88102
* __->__ #88045

This reverts commit bc64999b8382796199178cf480adf51512b5f139.

Check torch/_subclasses/meta_utils.py for "This is very tricky" for the bugfix explanation.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

**Comment by ezyang** [comment]
This is the bugfix diff from first time https://gist.github.com/ezyang/e2e8238802dc187ea1e02059d9c48599

**Review by kit1980 — APPROVED** [review]

**Comment by ezyang** [comment]
I undid the Unsupported custom function as apparently custom functions don't support symbolic shapes (we will need to fix that...)

**Review by Chillee — APPROVED** [review]

**Comment by ezyang** [comment]
Good news! I found the bug in meta converter that was causing the failures. So the patch bugfix has been replaced with a proper bugfix for meta converter, and extra asserts to catch any similar problems.

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -139,15 +138,17 @@ def tree_flatten_only(ty: Type[T], pytree: PyTree):
...
     constant_storage_mapping: Dict[StorageWeakRef, List[TensorWeakRef]]
 
     def __init__(self):
-        # FakeTensors store the FakeTensorMode which in turn stores a
-        # FakeTensor, so we need to hold a weak reference to the FakeTensor
-        # otherwise we would induce a circular reference
-        self.tensor_memo = weakref.WeakValueDictionary()
+        # In principle preserving views should be OK, but in practice
+        # AOTAutograd (or maybe autograd) seems to do the wrong thing.  See
+        # https://github.com/pytorch/torchdynamo/issues/1815
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:151]: Is this comment out of date now? https://github.com/pytorch/torchdynamo/issues/1815

Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:151]: yes, out of date

File: torch/_subclasses/meta_utils.py [hunk 2]
```diff
@@ -127,18 +170,31 @@ def del_ten():
...
         # Use a Weak Ref to s in order to not leak memory
         swr = StorageWeakRef(s)
         if swr not in self.storage_memo:
-            self.storage_memo[swr] = torch.empty(s.size(), dtype=s.dtype, device="meta")
+            self.storage_memo[swr] = (
+                callback(
+                    lambda: torch.empty(s.size(), dtype=torch.uint8, device="meta")
+                )
+                .storage()
+                .untyped()
```
Comment (bdhirsh) [inline torch/_subclasses/meta_utils.py:185]: is it important here that we have untyped storage in the storage memo, and not the dtype of the passed in tensor?

**Pull request closed without merging**